### PR TITLE
Resolve false negative for `used-before-assignment`  with TYPE_CHECKING guard

### LINKED ISSUE #8893: false negative: used-before-assignment w/ TYPE_CHECKING guard
### Bug description

Consider the following code:

```python
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import datetime

class Z:
    def something(self) -> None:
        z = datetime.datetime.now() # not imported at runtime
```

Pylint can and does flag the `z = ...` line as used-before-assignment. However, if we use `datetime` somewhere as a type annotation, say in

```python
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import datetime


def x() -> datetime.datetime: # this is good
    return datetime.datetime.now() # this should be bad


class Z:
    def something(self) -> None:
        z = datetime.datetime.now() # this should still be bad
```

pylint returns "10/10, would lint again". pylint should flag the `used-before-assignment` errors here, since this code will blow up at runtime when called.

### Command used

```shell
pylint --disable all --enable E <file>
```

### Pylint output

```shell
* without type annotation
% pylint --disable all --enable E xyz.py
************* Module xyz
xyz.py:15:12: E0601: Using variable 'datetime' before assignment (used-before-assignment)

------------------------------------------------------------------
Your code has been rated at 2.86/10 (previous run: 0.00/10, +2.86)

* with type annotation
% pylint --disable all --enable E xyz.py

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 0.00/10, +10.00)
```

### Expected behavior

consistent `used-before-assignment` lints.

### Pylint version

```shell
% pylint --version                      
pylint 2.17.3
astroid 2.15.4
Python 3.10.12 (main, Jun 12 2023, 08:08:55) [GCC 11.2.0]
```

### OS / Environment

Debian

### PR DESCRIPTION
**Pull request by zenlyj**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
|    | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Description

In the function responsible for filtering nodes defined under a `TYPE_CHECKING` block from consumption, usage scopes should only be updated when `used-before-assignment` is reported.

However, there is currently a misalignment between the `_report_unfound_name_definition()` and the filter function regarding when the error is raised. To resolve this, `_report_unfound_name_definition()` has been updated to return the check result, which is now passed to the filter function.

Additionally, the check has been extended to handle cases where a class is defined within an `if` block and used later.

Closes #8893

### CODE REVIEW

**Comment by zenlyj** [comment]
setting this to draft first, as i would like to check this more thoroughly. but feel free to discuss / give feedback!

**Comment by zenlyj** [comment]
@jacobtylerwalls can i get a review for this PR?

regarding the primer result, the sentry one seems legitimate. as an import is made under `TYPE_CHECKING`, then used for an `isinstance` check. the django ones should be reasonable reports as well? where a class is defined within an `if` block, but not within the associated `else` block, hence raising `possibly-used-before-assignment`

File: pylint/checkers/variables.py [hunk 1]
```diff
@@ -723,6 +723,8 @@ def _uncertain_nodes_if_tests(
                 name = other_node.name
             elif isinstance(other_node, (nodes.Import, nodes.ImportFrom)):
                 name = node.name
+            elif isinstance(other_node, nodes.ClassDef):
```
Comment (jacobtylerwalls) [inline pylint/checkers/variables.py:726]:
```suggestion
            elif isinstance(other_node, (nodes.FunctionDef, nodes.ClassDef)):
```

File: tests/functional/u/used/used_before_assignment_typing.py [hunk 2]
```diff
@@ -174,7 +174,7 @@ def defined_in_elif_branch(self) -> calendar.Calendar:  # [possibly-used-before-
     def defined_in_else_branch(self) -> urlopen:
         print(zoneinfo)  # [used-before-assignment]
         print(pprint())
```
Comment (jacobtylerwalls) [inline tests/functional/u/used/used_before_assignment_typing.py:176]:
```suggestion
        print(pprint())  # [used-before-assignment]
```

**Review by jacobtylerwalls — COMMENTED** [review]
Thank you! Just a small request to handle one more case.

**Review by jacobtylerwalls — DISMISSED** [review]

**Comment by jacobtylerwalls** [comment]
Any idea about the new Django message for a nonlocal?

**Comment by zenlyj** [comment]
seems that the message was raised due to a guarded function definition further down at line 1584. but from my understanding, this is a false positive? it should not flag as `condition` is provided as a function parameter in the enclosing function `_deferredSkip()`

**Comment by jacobtylerwalls** [comment]
I wonder if we need to explicitly handle nonlocals somewhere. If you don't have bandwidth then I guess we can back out your last change and punt this to a new issue so that we don't cause more false positives.

**Comment by zenlyj** [comment]
perhaps we should drop the latest commit and track this case with a separate issue? so that we can unblock #9964.

you may assign the issue to me and i'll find time to work on it

**Comment by jacobtylerwalls** [comment]
Very good--thank you!

**Comment by nickdrozd** [comment]
There is also the problem of `unused-import` false negatives:

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from math import pi  # unused-import false negative
```

Is it possible to fix that along the way?

**Comment by jacobtylerwalls** [comment]
> you may assign the issue to me and i'll find time to work on it

Again, thanks for this offer. I opened #10028

**Comment by zenlyj** [comment]
> > you may assign the issue to me and i'll find time to work on it
> 
> Again, thanks for this offer. I opened #10028

hey no worries, i dropped the latest commit so this should be good to go.



> There is also the problem of `unused-import` false negatives:
> 
> ```python
> from typing import TYPE_CHECKING
> 
> if TYPE_CHECKING:
>     from math import pi  # unused-import false negative
> ```
> 
> Is it possible to fix that along the way?

sure thing, i added it to #10028

**Review by Pierre-Sassoulas — APPROVED** [review]
The sentry primer looks legit.

**Pull request merged**